lib/metaheuristics/ga.ts

```diff
@@ -0,0 +1,119 @@
+
+    public selectionFn: 'Roulette' | 'Tournament' = 'Roulette'
+}
+
+export default class GA extends PopulationMetaheuristic<GAConfig> {
+    private convergence!: Convergence
+
+    private selection!: selectionFunctionMappings.Roulette | selectionFunctionMappings.Tournament
+
+    constructor(config?: GAConfig | undefined) {
+        if (config && config.selectionSize % 2 === 1) {
+            config.selectionSize++
+        }
+        if (config && config.selectionSize > config.populationSize) {
+            throw new Error('Selection size has to be smaller than population size')
+        }
+        super(Object.assign(new GAConfig(), config))
+        this.initializeDependencies()
+    }
+
+    private crossover(individual: Individual, parent1: Individual, parent2: Individual) {
+        const splitLocation = Random.randomInt(0, this.dimensions.length)
+        for (let j = 0; j < this.dimensions.length; j++) {
+            if (
+                (this.config.crossoverType === 'singleSplit' && splitLocation < j) ||
+                (this.config.crossoverType === 'randomPick' && Random.random(0, 1) < 0.5)
+            ) {
+                individual.position[j] = parent1.position[j]
+                individual.bestPosition[j] = parent1.position[j]
+            } else {
+                individual.position[j] = parent2.position[j]
+                individual.bestPosition[j] = parent2.position[j]
+            }
+        }
+    }
+
+    private mutation(individual: Individual) {
+        const mutate = (pos: number) => {
+            individual.position[pos] = Random.random(this.dimensions[pos].min, this.dimensions[pos].max)
+
+            if (individual.position[pos] > this.dimensions[pos].max) {
+                individual.position[pos] = this.dimensions[pos].max
+            }
+            if (individual.position[pos] < this.dimensions[pos].min) {
+                individual.position[pos] = this.dimensions[pos].min
+            }
+        }
+
+        if (this.config.mutationType === 'single') {
+            mutate(Random.seededRandomInt(0, this.dimensions.length))
+        } else {
+            for (let j = 0; j < this.dimensions.length; j++) {
+                const rand = Random.random(0, 1)
+                if (rand < this.config.mutationProbability) {
+                    mutate(j)
+                }
+            }
+        }
+    }
+
+    protected movePosition(individual: Individual, parent1: Individual, parent2: Individual) {
+        this.crossover(individual, parent1, parent2)
+        this.mutation(individual)
+    }
+
+    public step() {
+        // refactor
+        this.computeFitness()
+        const selected = this.selection.execute(this.individuals)
+        for (let i = 0; i < selected.length; i += 2) {
+            const parent1: Individual = selected[i]
+            const parent2: Individual = selected[i + 1]
+
+            const child = new Individual()
```

Comment:
we can change the config to accept property: numOffsprings instead of selected in order be more precise what is the result of selection